validator_client/src/beacon_node_fallback.rs

```diff
@@ -420,24 +467,49 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
     /// low quality responses. To route around this it's best to poll all connected beacon nodes.
     /// A previous implementation of this function polled only the unavailable BNs.
     pub async fn update_all_candidates(&self) {
-        let futures = self
-            .candidates
-            .iter()
-            .map(|candidate| {
-                candidate.refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
-            })
-            .collect::<Vec<_>>();
+        let candidates = self.candidates.read().await;
+        let mut futures = Vec::with_capacity(candidates.len());
+        let mut nodes = Vec::with_capacity(candidates.len());
+
+        for candidate in candidates.iter() {
+            futures.push(candidate.refresh_health(
+                &self.distance_tiers,
+                self.slot_clock.as_ref(),
+                &self.spec,
+                &self.log,
+            ));
+            nodes.push(candidate.beacon_node.to_string());
+        }
 
-        // run all updates concurrently and ignore errors
-        let _ = future::join_all(futures).await;
+        // Run all updates concurrently.
+        let future_results = future::join_all(futures).await;
+        let results = future_results.iter().zip(nodes);
+
+        for (result, node) in results {
+            if let Err(e) = result {
+                if *e != CandidateError::PreGenesis {
+                    warn!(
+                        self.log,
+                        "A connected beacon node errored during routine health check";
+                        "error" => ?e,
+                        "endpoint" => node,
+                    );
+                }
+            }
+        }
+
+        drop(candidates);
+
+        let mut candidates = self.candidates.write().await;
```

Comment:
is it possible to acquire the write lock after we get all the health details, and just before we update it?

I'm worried that these locking mechanisms would delay the requests, because a call to `first_success`, `broadcast` would have to wait for the iteration of querying all node health to finish before it can start

Reply:
^ This is actually not true, sorry!

Reply:
The scenario I'm concerned about is:

1. **Thread 1** calls `broadcast` for an endpoint that takes 5s due to a slow BN. This thread acquires the `candidates` read lock here for 5 s:
https://github.com/sigp/lighthouse/blob/405282764ec5a82b1c5bea95fc56f7282f81aa2c/validator_client/src/beacon_node_fallback.rs#L624
2. **Thread 2** calls `update_all_candidates` during this time, at the end of it it tries to acquire the `candidates` write lock, however this has to wait until (1) releases the read lock
https://github.com/sigp/lighthouse/blob/405282764ec5a82b1c5bea95fc56f7282f81aa2c/validator_client/src/beacon_node_fallback.rs#L503
3. **Thread 3** calls `broadcast` on another endpoint, it needs to acquire the `candidates` read lock, but because the write lock from (2) is pending, it has to wait
4. After 5 seconds, (1) releases the read lock, allowing (2) to acquire the write lock, this write should be fast, so it immediately releases it
5. Now (3) can finally acquire the read lock and make the request

Not sure if my understanding of the code is correct though. @chong-he and I paired together a bit to work on a way to test this, but it's still a WIP, we've pushed a branch here
https://github.com/jimmygchen/lighthouse/tree/vc-fallback-test

Reply:
Yep this is definitely a concern. We actually hold locks across awaits in a bunch of places. I don't think there was any deadlock risk but it's a big no-no and can lead to the situation you're describing. 

My solution has been to tweak things and clone the `BeaconNodeHttpClient` where necessary. This is a relatively cheap clone (although the presence of `timeouts` is frustrating, and I wonder if those values can be moved elsewhere. I actually added them a while back and now I'm suffering the consequences :upside_down_face: )

I believe now, we will always drop the lock before awaiting the futures. This does come at a cost, such as having to clone or reacquire the lock in some cases, but I think this trade off is both acceptable and necessary. Let me know what you think!

Reply:
Yeah sounds reasonable to me and a practical solution I think. We can work on the potential improvements to timeouts and cloning in a separate PRs if necessary. This looks good to me now.